File: 2015/09.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "set"

graph  = Hash.new { |h, k| h[k] = Hash.new }
places = Set.new
min    = Float::INFINITY
max    = 0

File.open("./input/09.txt").each_line do |line|
  /(?<l0>\w+) to (?<l1>\w+) = (?<d>\d+)/ =~ line
  graph[l0][l1] = d.to_i
  graph[l1][l0] = d.to_i
  places << l0
  places << l1
end

places.to_a.permutation.each do |path|
  tmp = 0
  (0...path.length - 1).each do |i|
    dist = graph[path[i]][path[i + 1]]
    tmp += dist
  end
  min = [min, tmp].min
  max = [max, tmp].max
end

puts("
Day 09
======
✮: #{min}
★: #{max}")

File: 2016/09.py
#!/usr/bin/env python3
import regex as re


def decompress(line, recurv=False):
    size = 0
    i = 0
    while i < len(line):
        while i < len(line) and line[i] != '(':
            i += 1
            size += 1
        if i == len(line):
            break

        i += 1

        sublen = 0
        cycles = 0

        while line[i] != 'x':
            sublen *= 10
            sublen += int(line[i])
            i += 1
        i += 1

        while line[i] != ')':
            cycles *= 10
            cycles += int(line[i])
            i += 1
        i += 1

        if recurv:
            size += cycles * decompress(line[i:i + sublen], recurv)
        else:
            size += cycles * sublen
        i += sublen

    return size


file = re.sub("\s+", '', open("input/09.txt").read())
print(decompress(file))
print(decompress(file, recurv=True))
